plutus-wallet-api/ledger/Ledger/Oracle.hs

```diff
@@ -0,0 +1,149 @@
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE NamedFieldPuns    #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# OPTIONS_GHC -Wno-simplifiable-class-constraints #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# OPTIONS_GHC -fno-strictness #-}
+{-# OPTIONS_GHC -fno-specialise #-}
+{-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
+module Ledger.Oracle(
+  -- $oracles
+  Observation(..)
+  , SignedMessage(..)
+  , SignedMessageCheckError(..)
+  , checkSignature
+  , checkHashOnChain
+  , checkHashOffChain
+  , signMessage
+  ) where
+
+import qualified Data.ByteString.Lazy      as BSL
+import           GHC.Generics              (Generic)
+
+import           Language.PlutusTx
+import           Language.PlutusTx.Prelude
+
+import           Ledger.Crypto             (PrivateKey, PubKey (..), Signature (..))
+import qualified Ledger.Crypto             as Crypto
+import           Ledger.Scripts            (DataValue (..), DataValueHash (..))
+import qualified Ledger.Scripts            as Scripts
+import           Ledger.Slot               (Slot)
+import           Ledger.Validation         (PendingTx')
+import qualified Ledger.Validation         as V
+import           LedgerBytes
+
+import qualified Prelude                   as Haskell
+
+-- $oracles
+-- This module provides a way to verify signed messages, and a type for
+-- observations (for example, the price of a commodity on a given date).
+-- Together, the two can be used to implement trusted oracles:
+-- * The oracle observes a value, for example 'Price' and constructs a value
+--   'o = Observation Price'.
+-- * The oracle then calls 'signMessage o pk' with its own private key to
+--   produce a 'SignedMessage (Observation Price)'.
+-- * The signed message is passed to the contract as the redeemer of some
+--   unspent output. IMPORTANT: The redeeming transaction must include the
+--   message 'o' as a data value. This is because we can't hash anything in
+--   on-chain code, and therefore have to rely on the node to do it for us
+--   via the pending transaction's map of data value hashes to data values.
+-- * The contract then calls 'checkSignature' to check the signature, and
+--   'checkHashOnChain' to ensure that the signed hash is really the hash of
+--   the data value.
+
+-- | A value that was observed at a specific point in time
+data Observation a = Observation
+    { obsValue :: a
+    -- ^ The value
+    , obsSlot  :: Slot
```

Comment:
I think 'Slot' is fine here since this is an observation about the past. Say, the oil price on March 3 1997. The observation is not invalidated by new observations on different date.